File: 002_ValidAnagram.cpp
// Topic: Valid Anagram
// Description: Determine if two strings have identical character frequencies.
// Approach: Count characters using fixed-size array (assuming ASCII lowercase) or unordered_map for
// general case.

#include <bits/stdc++.h>
using namespace std;

class Solution
{
   public:
    bool isAnagram(const string& s, const string& t)
    {
        if (s.size() != t.size())
            return false;
        array<int, 26> cnt{};
        for (size_t i = 0; i < s.size(); ++i)
        {
            ++cnt[s[i] - 'a'];
            --cnt[t[i] - 'a'];
        }
        for (int x : cnt)
            if (x)
                return false;
        return true;
    }
};

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    vector<pair<string, string>> tests = {{"anagram", "nagaram"}, {"rat", "car"}};
    for (size_t i = 0; i < tests.size(); ++i)
    {
        auto& p = tests[i];
        cout << "Test " << (i + 1) << ": (" << p.first << "," << p.second << ") -> "
             << (Solution().isAnagram(p.first, p.second) ? "true" : "false") << "\n";
    }
}

/*
Time Complexity: O(n)
Space Complexity: O(1) (26-size array)
*/

File: 002_ValidAnagram.py
# Topic: Valid Anagram
# Description: Determine if two strings have identical character frequencies.
# Approach: Count characters using a dictionary or Counter.

from collections import Counter


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        cnt = [0] * 26
        for i in range(len(s)):
            cnt[ord(s[i]) - ord('a')] += 1
            cnt[ord(t[i]) - ord('a')] -= 1
        return all(x == 0 for x in cnt)


if __name__ == "__main__":
    tests = [("anagram", "nagaram"), ("rat", "car")]
    sol = Solution()
    for i, (s, t) in enumerate(tests):
        result = sol.isAnagram(s, t)
        print(f"Test {i + 1}: ({s},{t}) -> {result}")

"""
Time Complexity: O(n)
Space Complexity: O(1) (26-size array)
"""
